plugins/sinks/image/image.go

```diff
@@ -0,0 +1,177 @@
+			}
+			m.format = i
+		}
+	}
+
+	m.path = "image"
+	if i, ok := props["path"]; ok {
+		if i, ok := i.(string); ok {
+			m.path = i
+		}
+	}
+
+	m.maxAge = 72
+	if i, ok := props["maxAge"]; ok {
+		if i, ok := i.(int); ok {
+			m.maxAge = i
+		}
+	}
+	m.maxCount = 1000
+	if i, ok := props["maxCount"]; ok {
+		if i, ok := i.(int); ok {
+			m.maxCount = i
+		}
+	}
+	return nil
+}
+
+func (m *imageSink) Open(ctx api.StreamContext) error {
+	logger := ctx.GetLogger()
+	logger.Debug("Opening image sink")
+
+	if _, err := os.Stat(m.path); os.IsNotExist(err) {
+		if err := os.MkdirAll(m.path, os.ModePerm); nil != err {
+			return fmt.Errorf("fail to open image sink for %v", err)
+		}
+	}
+
+	t := time.NewTicker(time.Duration(3) * time.Minute)
+	exeCtx, cancel := ctx.WithCancel()
+	m.cancel = cancel
+	go func() {
+		defer t.Stop()
+		for {
+			select {
+			case <-t.C:
+				m.delFile(logger)
+			case <-exeCtx.Done():
+				logger.Info("image sink done")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *imageSink) delFile(logger api.Logger) error {
+	files, err := ioutil.ReadDir(m.path)
+	if nil != err || 0 == len(files) {
+		return err
+	}
+
+	pos := m.maxCount
+	delTime := time.Now().Add(time.Duration(0-m.maxAge) * time.Hour)
+	for i := 0; i < len(files); i++ {
+		for j := i + 1; j < len(files); j++ {
+			if files[i].ModTime().Before(files[j].ModTime()) {
+				files[i], files[j] = files[j], files[i]
+			}
+		}
+		if files[i].ModTime().Before(delTime) && i < pos {
+			pos = i
+			break
+		}
+	}
+
+	for i := pos; i < len(files); i++ {
+		fname := files[i].Name()
+		if strings.HasSuffix(fname, m.format) {
+			fpath := filepath.Join(m.path, fname)
+			os.Remove(fpath)
+		}
+	}
+	return nil
+}
+
+func (m *imageSink) getSuffix() string {
+	now := time.Now()
+	year, month, day := now.Date()
+	hour, minute, second := now.Clock()
+	nsecond := now.Nanosecond()
+	return fmt.Sprintf(`%d-%d-%d_%d-%d-%d-%d`, year, month, day, hour, minute, second, nsecond)
+}
+
+func (m *imageSink) saveFile(msg []map[string][]byte) error {
+	for _, images := range msg {
+		for k, v := range images {
+			reader := bytes.NewReader(v)
+			suffix := m.getSuffix()
+			fname := fmt.Sprintf(`%s%s.%s`, k, suffix, m.format)
+			fpath := filepath.Join(m.path, fname)
+			fp, err := os.Create(fpath)
+			if nil != err {
+				return err
+			}
+			defer fp.Close()
+			if "png" == m.format {
+				if img, err := png.Decode(reader); nil != err {
+					return err
+				} else {
+					return png.Encode(fp, img)
+				}
+			} else if "jpeg" == m.format {
+				if img, err := jpeg.Decode(reader); nil != err {
+					return err
+				} else {
+					return jpeg.Encode(fp, img, nil)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (m *imageSink) Collect(ctx api.StreamContext, item interface{}) error {
+	logger := ctx.GetLogger()
+	if v, ok := item.([]byte); ok {
+		var msg []map[string][]byte
+		if err := json.Unmarshal(v, &msg); nil != err {
+			return err
```

Comment:
Suggest to give a clearer error message for the user. The sink only accepts bytea field, other types are not supported.